File: java/src/com/alofmethbin/jadawin/elements/Gallery.java
package com.alofmethbin.jadawin.elements;

import com.alofmethbin.jadawin.Article;
import java.io.IOException;
import java.util.ArrayList;
import java.util.List;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

public class Gallery extends Element {
    private List<Text>  labels   = new ArrayList<>();
    private List<Image> images   = new ArrayList<>();
    private static Pattern crack = Pattern.compile( "^(\\S+)\\s+(.+)$");
    
    public Gallery( Article article, List<String> lines) throws IOException {
        super( article);
        
        for (String line: lines) {
            Matcher m = crack.matcher(line);
            if ( m.find() ) {
                Image image = new Image( article, m.group(1));
                Text text = new Text( article, m.group(2));
                if (! (image.hasError() || text.hasError())) {
                    images.add( image);
                    labels.add( text);
                }
            }
        }
        
        if ( images.isEmpty() ) {
            error( "Gallery empty");
        }
    }

    @Override
    public Image getImage() {
        return images.get(0);
    }

    @Override
    public boolean isOverlay() {
        return true;
    }

    @Override
    public int lineCount() {
        return 5 * ((images.size() + 7) / 8);
    }

    @Override
    public void prepare() {
        for (Text label: labels) {
            label.prepare();
        }
    }
}

File: ruby/elements/gallery.rb
require_relative 'image'

module Elements
  class Gallery < Element
    include Utils

    def initialize( compiler, article, lines)
      super( article)
      @images, @labels = [], []
      lines.each do |line|
        if m = /^(\S+)\s+(.*)$/.match( line)
          @images << Image.new( compiler, article, [m[1]])
          @labels << Text.new( compiler, article, [m[2]])
        elsif line.strip != ''
          article.error( 'No label for ' + line.strip)
        end
      end
    end

    def image
      @images[0]
    end

    def line_count
      5 * ((@images.size + 7) / 8).to_i
    end

    def prepare( compiler, article, parents, younger)
      @labels.each do |text|
        text.prepare( compiler, article, parents, younger)
      end
    end

    def to_data( compiler, article)
      icon_dims = get_scaled_dims( compiler.dimensions( 'icon'), @images)

      entries = []
      @images.each_index do |i|
        entries << {'tag'     => [@labels[i].to_data_single_paragraph( compiler, article)],
                    'overlay' => @images[i].overlay( compiler, article),
                    'details' => @images[i].details( compiler, article, icon_dims, :prepare_thumbnail)}
      end

      {'type'   => 'gallery',
       'dims'   => icon_dims,
       'images' => entries}
    end
  end
 end